### PR TITLE
ASEMD refactor

### PR DESCRIPTION
**Pull request by Tetracarbonylnickel**
This PR implements a overhaul of the ASEMD Node and a couple of features related to it.

- ASEMD Node
    - steps is now the total number of MD steps
    - sampling_rate is the number that defines after how many MD steps a structure is loaded to the cache
    - progress bar now gives proper it/s
    - progress bar also shows how many fs the md is running
    - init_velocities now zntrack.zn.deps() and velocities_cache zntrack.zn.outs(), with that MD simulations can now be continued within one project
    - constraints now can be set in similar way as the checker and the thermostat
    - metrics is now written every sampling_rate iteration
    - having multiple ThresholdChecker is now possible
    
- FixedShpereMD is now a constraint Node (similar than checker)

- Checker are overhauled

- EMT calculator can now write stress to hdf5 files

- addresses issue #158 #133. In principle also #142 because checker write the relevant metrics into metrics_cache which is zn.plots(). Maybe extra hook for tracking certain metrics during a MD? Is very fast implemented.

### CODE REVIEW

File: ipsuite/analysis/model/dynamics_checks.py [hunk 1]
```diff
@@ -179,19 +224,30 @@ def check(self, atoms) -> bool:
...
                 f"StandardDeviationCheck for '{self.value}' triggered by"
-                f" '{self.values[-1]:.3f}' for '{np.mean(self.values):.3f} +-"
-                f" {np.std(self.values):.3f}' and max value '{self.max_value}'"
+                # dependent on self.value self.values can be an ndarray
+                # then this status gives an error message np.max(self.values)
+                # should be the best option here. but I'm not sure if it
+                # is what you intended.
+                f" '{np.max(self.values[-1]):.3f}' for '{mean:.3f} +-"
+                f" {std:.3f}' and max value '{self.max_value}'"
             )
```
Comment (Tetracarbonylnickel) [inline ipsuite/analysis/model/dynamics_checks.py:242]: I was not sure with that. Maybe you can give me some input if this is how you intend it.

File: ipsuite/analysis/model/dynamics_checks.py [hunk 2]
```diff
@@ -23,7 +26,17 @@ def check(self, atoms: ase.Atoms) -> bool:
...
+            self.status = (
+                "NaN check failed: last iterationpositions energy or forces = NaN"
+            )
+            return True
+        else:
+            self.status = "No NaN occurred"
+            return False
+
+    def __str__(self):
+        return self.status
```
Comment (PythonFZ) [inline ipsuite/analysis/model/dynamics_checks.py]: Could this be part of the parent class?

File: ipsuite/analysis/model/dynamics_checks.py (hunk 1 above)
Comment (PythonFZ) [reply ipsuite/analysis/model/dynamics_checks.py:230]: I guess for non-skalar values we can use the maximum value. Would this also require `if np.max(distance) > self.max_std * std` ?

File: ipsuite/calculators/ase_md.py [hunk 3]
```diff
@@ -311,18 +341,21 @@ class ASEMD(base.ProcessSingleAtom):
...
     steps: int
-        number of steps to simulate
+        total number of steps of the simulation
     sampling_rate: int
-        number of sample runs
+        number defines after how many md steps a structure
+        is loaded to the cache
     metrics_dict:
-        saved total energy and all metrics from the check nodes
+        saved total energy and metrics from the check nodes
```
Comment (PythonFZ) [inline ipsuite/calculators/ase_md.py:358]: `metrics_dict` might be the wrong name for this data. Maybe we could rename it to `thermo` like in lammps https://docs.lammps.org/thermo.html

Comment (PythonFZ) [reply ipsuite/calculators/ase_md.py:358]: Could we also introduce `thermo_interval: int` to update this data in a different interval than dumping the atoms.
The reason for this is, that with e.g. 100_000 we have > 200_000 datapoints in the csv file and plotting them can be very slow so we might only want the thermo data every 1000 steps while we dump the trajectories every step?

**Review by PythonFZ — CHANGES_REQUESTED** [review]
Thanks for the MD fix. This was required and looks much, much better now! Only some small comments and maybe one addition.

File: ipsuite/analysis/model/dynamics_checks.py (hunk 2 above)
Comment (PythonFZ) [reply ipsuite/analysis/model/dynamics_checks.py]: fixed in 51f5b73

File: ipsuite/calculators/ase_md.py (hunk 3 above)
Comment (PythonFZ) [reply ipsuite/calculators/ase_md.py:358]: To simplify things I'd suggest:
- default for `thermo_interval: int = zn.params(None)` and make ist the same as `sampling_interval`
- only allow `thermo_interval` to be a multiple of `sampling_interval`. We can check that in the `_post_init_`.

Comment (Tetracarbonylnickel) [reply ipsuite/calculators/ase_md.py:358]: Isn't it better to have the data and than use every nth datapoint for plotting? Sometimes you may want to zoom in some regions or want to see how smooth the kurve is?

Comment (PythonFZ) [reply ipsuite/calculators/ase_md.py:358]: Let's leave it as is for now and maybe do that in another PR. Maybe there is also a way to do this on the DVC side.

File: ipsuite/analysis/model/dynamics_checks.py (hunk 1 above)
Comment (Tetracarbonylnickel) [reply ipsuite/analysis/model/dynamics_checks.py]: Yes you are right and it is implemented now.

**Pull request merged**